### PR TITLE
feat: add cross device sync

### PR DESCRIPTION
**Pull request by kaiserbh**
ref: [GitHub - Tachiyomi Pull Request #9790](https://github.com/tachiyomiorg/tachiyomi/pull/9790)

Currently, there are two services available: [Syncyomi](https://github.com/SyncYomi/SyncYomi), which is self-hosted, and Google Drive.

Once merged you will need to create OAuth 2.0 Client IDs in google console then replace the `app/src/main/assets/client_secrets.json`

image

Once all of that is done just need to set the app to `publish` instead of `testing` that way everyone can access it otherwise limited to 100 users we add for testing, we don't need to submit the app for approval since we are not using sensitive scopes. More info [here](https://support.google.com/cloud/answer/13463073?authuser=1&visit_id=638407747036157208-1875300124&rd=1)

### OAuth Consent Screen
image

### Scopes
image

image
image
image
image
image

### CODE REVIEW

**Comment by AntsyLich** [comment]
Initially I'll do reviews on small independent sections and ask you to make separate pr for them e.g. the db change you did which you can go ahead and make a pr for

**Comment by KaiserBh** [comment]
> Initially I'll do reviews on small independent sections and ask you to make separate pr for them e.g. the db change you did which you can go ahead and make a pr for

#113

File: domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        fun values() = listOf(
+            NONE,
+            SYNC_ON_CHAPTER_READ,
+            SYNC_ON_CHAPTER_OPEN,
+            SYNC_ON_APP_START,
+            SYNC_ON_APP_RESUME,
+            SYNC_ON_LIBRARY_UPDATE,
+        )
+    }
```
Comment (AntsyLich) [inline domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt:28]: why not use an enum?

File: domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+data class SyncOptions(
+    val libraryEntries: Boolean = true,
+    val categories: Boolean = true,
+    val chapters: Boolean = true,
+    val tracking: Boolean = true,
+    val history: Boolean = true,
+    val appSettings: Boolean = true,
+    val sourceSettings: Boolean = true,
+    val privateSettings: Boolean = false,
+)
```
Comment (AntsyLich) [inline domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt:100]: Move into separate `model` package. Rename to `SyncSettings`?

File: domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            SYNC_ON_CHAPTER_READ,
+            SYNC_ON_CHAPTER_OPEN,
+            SYNC_ON_APP_START,
+            SYNC_ON_APP_RESUME,
+            SYNC_ON_LIBRARY_UPDATE,
+        )
+    }
+
+    fun syncHost() = preferenceStore.getString("sync_host", "https://sync.tachiyomi.org")
+    fun syncAPIKey() = preferenceStore.getString("sync_api_key", "")
```
Comment (AntsyLich) [inline domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt:31]:
```suggestion
    fun clientHost() = preferenceStore.getString("sync_client_host", "127.0.0.1:69420")
    fun clientAPIKey() = preferenceStore.getString("sync_client_api_key", "")
```

File: domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        var uniqueID = uniqueIDPreference.get()
+        if (uniqueID.isBlank()) {
+            uniqueID = UUID.randomUUID().toString()
+            uniqueIDPreference.set(uniqueID)
+        }
+
+        return uniqueID
+    }
+
+    fun syncFlags() = preferenceStore.getInt("sync_flags", Flags.Defaults)
```
Comment (AntsyLich) [inline domain/src/main/java/tachiyomi/domain/sync/SyncPreferences.kt:60]:
```suggestion
    fun autoSyncAfter() = preferenceStore.getInt("auto_sync_after", Flags.Never)
```

**Review by AntsyLich — CHANGES_REQUESTED** [review]
Make a separate PR for this file and also maybe prefix all of them with `__SYNC__`?

**Comment by KaiserBh** [comment]
> Make a separate PR for this file and also maybe prefix all of them with `__SYNC__`?

#149 Let me know on that branch if I need to change anything else.

**Comment by LightDestory** [comment]
Build fails?

**Comment by KaiserBh** [comment]
> Build fails?

Hmm the build is okay it's just detekt errors, which is just too strict at the moment, to name a few:

`Class 'SyncPreferences' with '13' functions detected. Defined threshold inside classes is set to '11' [TooManyFunctions]`

`/home/runner/work/mihon/mihon/app/src/main/java/eu/kanade/tachiyomi/data/sync/SyncManager.kt:371:43: This expression contains a magic number. Consider defining it to a well named constant. [MagicNumber]`

`/home/runner/work/mihon/mihon/app/src/main/java/eu/kanade/tachiyomi/data/sync/SyncManager.kt:372:44: This expression contains a magic number. Consider defining it to a well named constant. [MagicNumber]`

**Pull request closed without merging**